From f1bc7172c3fb4b1c45cd84f09ad3260ecbf06dac Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 11:40:39 -0400
Subject: [net] ethtool: Return immediately on error in ethtool_copy_validate_indir()

Message-id: <1410176446-10440-4-git-send-email-ivecera@redhat.com>
Patchwork-id: 91652
O-Subject: [RHEL7.1 PATCH 03/10] ethtool: Return immediately on error in ethtool_copy_validate_indir()
Bugzilla: 1138823
RH-Acked-by: Michal Schmidt <mschmidt@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: #1138823

Upstream commit(s):
commit fb95cd8d1473b1cc90eccbd6a30641f3851c8506
Author: Ben Hutchings <ben@decadent.org.uk>
Date:   Thu May 15 00:46:45 2014 +0100

    ethtool: Return immediately on error in ethtool_copy_validate_indir()

    We must return -EFAULT immediately rather than continuing into
    the loop.

    Similarly, we may as well return -EINVAL directly.

    Signed-off-by: Ben Hutchings <ben@decadent.org.uk>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/ethtool.c b/net/core/ethtool.c
index cc4f9fa..8e3b54a 100644
--- a/net/core/ethtool.c
+++ b/net/core/ethtool.c
@@ -550,19 +550,17 @@ static int ethtool_copy_validate_indir(u32 *indir, void __user *useraddr,
      struct ethtool_rxnfc *rx_rings,
      u32 size)
 {
- int ret = 0, i;
+ int i;
 
  if (copy_from_user(indir, useraddr, size * sizeof(indir[0])))
-  ret = -EFAULT;
+  return -EFAULT;
 
  /* Validate ring indices */
- for (i = 0; i < size; i++) {
-  if (indir[i] >= rx_rings->data) {
-   ret = -EINVAL;
-   break;
-  }
- }
- return ret;
+ for (i = 0; i < size; i++)
+  if (indir[i] >= rx_rings->data)
+   return -EINVAL;
+
+ return 0;
 }
 
 static noinline_for_stack int ethtool_get_rxfh_indir(struct net_device *dev,
-- 
1.7.1